src/Stratis.Features.Unity3dApi/NFTTransferIndexer.cs

```diff
@@ -0,0 +1,239 @@
+using System.Threading;
+using System.Threading.Tasks;
+using LiteDB;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Newtonsoft.Json;
+using Stratis.Bitcoin.AsyncWork;
+using Stratis.Bitcoin.Configuration;
+using Stratis.Bitcoin.Features.BlockStore.AddressIndexing;
+using Stratis.Bitcoin.Features.SmartContracts.Models;
+using Stratis.Bitcoin.Features.SmartContracts.Wallet;
+using FileMode = LiteDB.FileMode;
+
+namespace Stratis.Features.Unity3dApi
+{
+    public interface INFTTransferIndexer : IDisposable
+    {
+        /// <summary>Initialized NFT indexer.</summary>
+        void Initialize();
+
+        /// <summary>Adds NFT contract to watch list. Only contracts from the watch list are being indexed.</summary>
+        void WatchNFTContract(string contractAddress);
+
+        /// <summary>Provides a list of all nft contract addresses that are being tracked.</summary>
+        List<string> GetWatchedNFTContracts();
+
+        /// <summary>Provides collection of NFT ids that belong to provided user's address for watched contracts.</summary>
+        OwnedNFTsModel GetOwnedNFTs(string address);
+
+        /// <summary>Returns collection of all users that own nft.</summary>
+        NFTContractModel GetAllNFTOwnersByContractAddress(string contractAddress);
+    }
+
+    /// <summary>This component maps addresses to NFT Ids they own.</summary>
+    public class NFTTransferIndexer : INFTTransferIndexer
+    {
+        public ChainedHeader IndexerTip { get; private set; }
+
+        private const string DatabaseFilename = "NFTTransferIndexer.litedb";
+        private const string DbOwnedNFTsKey = "OwnedNfts";
+        private const int SyncBufferBlocks = 50;
+        
+        private readonly DataFolder dataFolder;
+        private readonly ILogger logger;
+        private readonly ChainIndexer chainIndexer;
+        private readonly IAsyncProvider asyncProvider;
+        private readonly ISmartContractTransactionService smartContractTransactionService;
+
+        private LiteDatabase db;
+        private LiteCollection<NFTContractModel> NFTContractCollection;
+        private CancellationTokenSource cancellation;
+        private Task indexingTask;
+        
+        public NFTTransferIndexer(DataFolder dataFolder, ILoggerFactory loggerFactory, IAsyncProvider asyncProvider, ChainIndexer chainIndexer, ISmartContractTransactionService smartContractTransactionService)
+        {
+            this.dataFolder = dataFolder;
+            this.cancellation = new CancellationTokenSource();
+            this.asyncProvider = asyncProvider;
+            this.chainIndexer = chainIndexer;
+            this.smartContractTransactionService = smartContractTransactionService;
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+        }
+
+        /// <inheritdoc />
+        public void Initialize()
+        {
+            string dbPath = Path.Combine(this.dataFolder.RootPath, DatabaseFilename);
```

Comment:
Since this is public maybe add a check that its not initialized more than once. E.g.:

```
public void Initialize()
{
    if (this.db != null)
      throw something.
    
      // Do initialization stuff.    
     ...
}
```

Reply:
Sure, done